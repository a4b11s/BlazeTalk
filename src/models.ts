export interface IUser {
	displayName: string;
	email: string;
	uid: string;
	photoURL: string;
}

export interface IChat {
	lastMessage: string;
	uid: string;
	name: string;
	chatAvatar: string;
}

export interface IMessage {
	chatUid: string;
	uid: string;
	author: {
		uid: string;
	};
	text: string;
	images?: Array<string>;
}
