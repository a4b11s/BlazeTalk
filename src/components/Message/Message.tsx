import React, { useEffect, useRef } from 'react';

import { Avatar, Box, Paper, Typography } from '@mui/material';

interface IProps {
	author: {
		displayName: string;
		photoURL: string;
	};
	message: { text: string };
	isAuthor: boolean;
	isScrollTo: boolean;
}

const Message = (props: IProps) => {
	const { message, isAuthor, author, isScrollTo } = props;
	const { text } = message;
	const ref = useRef<HTMLDivElement>(null);

	useEffect(() => {
		if (ref.current && isScrollTo) {
			ref.current.scrollIntoView();
		}
	}, [ref, isScrollTo]);

	const wrapperStyles = {
		borderRadius: `30px 30px ${isAuthor ? '0 30px' : '30px 0'};`,
		alignSelf: isAuthor ? 'flex-end' : 'flex-start',
		width: '35%',
		padding: 2,
		m: 3,
	};
	if (!message.text.trim()) return null;
	if (!author.displayName) return null;

	return (
		<Paper data-testid="message-wrapper" ref={ref} sx={wrapperStyles}>
			<Box display="flex" alignItems="center">
				<Avatar src={author.photoURL} />
				<Typography ml={1} color={'secondary.main'} fontWeight={500} variant="h6">
					{author.displayName}
				</Typography>
			</Box>
			<Typography data-testid="message-text" padding={1} variant="body1">
				{text}
			</Typography>
		</Paper>
	);
};

export default Message;
