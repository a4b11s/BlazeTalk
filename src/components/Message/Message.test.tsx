import React from 'react';

import { render, screen } from '@testing-library/react';

import { IMessage, IUser } from '../../models';

import Message from './Message';

const testMessage: IMessage = {
	text: 'TEST MESSAGES',
	uid: 'TEST UID',
	chatUid: 'TEST CHAT UID',
	author: {
		uid: 'TEST USER UID',
	},
};
const testUser: IUser = {
	uid: 'TEST USER UID',
	email: 'TEST EMAIL',
	photoURL: 'TEST PHOTO URL',
	displayName: 'TEST DISPLAY NAME',
};
const IncorrectTestUser: IUser = {
	uid: '',
	email: '',
	photoURL: '',
	displayName: '',
};

describe('Message component', () => {
	describe('should be render with correct props', () => {
		test('message should be rendered', () => {
			render(
				<Message
					message={testMessage}
					author={testUser}
					isAuthor={true}
					isLast={false}
				/>
			);
			expect(screen.getByText(testMessage.text)).toBeInTheDocument();
		});
		test('author name should be rendered', () => {
			render(
				<Message
					message={testMessage}
					author={testUser}
					isAuthor={true}
					isLast={false}
				/>
			);
			expect(screen.getByText(testUser.displayName)).toBeInTheDocument();
		});
	});
	describe('should not be error without correct props', () => {
		describe('should not be render without correct messages', () => {
			test('message is empty', () => {
				render(
					<Message
						message={{ ...testMessage, text: '' }}
						author={testUser}
						isAuthor={true}
						isLast={false}
					/>
				);
				expect(screen.queryByTestId('message-wrapper')).toBeNull();
				expect(screen.queryByTestId('message-text')).toBeNull();
			});
			test('message is white space', () => {
				render(
					<Message
						message={{ ...testMessage, text: ' ' }}
						author={testUser}
						isAuthor={true}
						isLast={false}
					/>
				);
				expect(screen.queryByTestId('message-wrapper')).toBeNull();
				expect(screen.queryByTestId('message-text')).toBeNull();
			});
			test('message is tab', () => {
				render(
					<Message
						message={{ ...testMessage, text: '	' }}
						author={testUser}
						isAuthor={true}
						isLast={false}
					/>
				);
				expect(screen.queryByTestId('message-wrapper')).toBeNull();
				expect(screen.queryByTestId('message-text')).toBeNull();
			});
		});
		test('should not be render without correct user', () => {
			render(
				<Message
					message={testMessage}
					author={IncorrectTestUser}
					isAuthor={true}
					isLast={false}
				/>
			);
			expect(screen.queryByTestId('message-wrapper')).toBeNull();
		});
	});
	test('should to match snapshot with correct props', () => {
		const view = render(
			<Message
				message={testMessage}
				author={testUser}
				isAuthor={true}
				isLast={false}
			/>
		);
		expect(view).toMatchSnapshot();
	});
});
