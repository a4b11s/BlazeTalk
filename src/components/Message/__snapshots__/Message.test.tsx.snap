// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component should to match snapshot with correct props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-17atwvm-MuiPaper-root"
        data-testid="message-wrapper"
      >
        <div
          class="MuiBox-root css-70qvj9"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          >
            <img
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="TEST PHOTO URL"
            />
          </div>
          <h6
            class="MuiTypography-root MuiTypography-h6 css-1cq3f58-MuiTypography-root"
          >
            TEST DISPLAY NAME
          </h6>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 css-130dmu7-MuiTypography-root"
          data-testid="message-text"
        >
          TEST MESSAGES
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-17atwvm-MuiPaper-root"
      data-testid="message-wrapper"
    >
      <div
        class="MuiBox-root css-70qvj9"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
        >
          <img
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="TEST PHOTO URL"
          />
        </div>
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1cq3f58-MuiTypography-root"
        >
          TEST DISPLAY NAME
        </h6>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-130dmu7-MuiTypography-root"
        data-testid="message-text"
      >
        TEST MESSAGES
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
